egret/model_library/unit_commitment/fuel_supply.py

```diff
@@ -0,0 +1,156 @@
+#  ___________________________________________________________________________
+#
+#  EGRET: Electrical Grid Research and Engineering Tools
+#  Copyright 2019 National Technology & Engineering Solutions of Sandia, LLC
+#  (NTESS). Under the terms of Contract DE-NA0003525 with NTESS, the U.S.
+#  Government retains certain rights in this software.
+#  This software is distributed under the Revised BSD License.
+#  ___________________________________________________________________________
+
+## file for all the ancillary services
+from pyomo.environ import *
+import math
+
+from .uc_utils import add_model_attr, build_uc_time_mapping
+from .status_vars import _is_relaxed
+
+## For now, this function will depend on having one of the production costs with
+## PiecewiseProduction variables, and start-up costs which use auxillary variables
+@add_model_attr('fuel_supply', requires = {'data_loader': None,
```

Comment:
Not for this initial implementation, but I think we should collectively start thinking of ways to tag model components at the point of definition, e.g., when defining CA_production_costs, label it (somehow) as satisfying the requirement "production_costs". This of course assumes the set of requirements is relatively static. This should help information-hiding. And maintenance, e.g., I could see having the dictionaries defined here being needed in other new components (e.g., dual fuel), which would required duplicate definitions.

Reply:
The (poorly named) `@add_model_attr` does some of this. If the requirements check passes, it sets `model.fuel_supply = 'fuel_supply_model'`. The requirements check iterates over the `requires` dictionary. First, it looks for the keys to be an attribute on the model. If the value for that key is `None`, it does no check on the attribute. If the value for that key is not none, it iterates over the list and checks to see if one of the list items is set for that attribute. If it can't find the key or the attribute for that key is set incorrectly, it throws an exception.

In this way, if you attempt to call this `fuel_supply_model` function before some of the others, you'll get an error before we get to the body of the function. So my comment here is more describing what the decorator below it is checking.